### PR TITLE
fix(persist): block throwing exception when storage is polluted

### PR DESCRIPTION
**Pull request by happyhj**
### tests
* Add suit of unit tests to test native host, not mocked environment.
* Add unit test for check that persist prevents throwing exception when storage is polluted.
* Change Array.forEach to jQuery.each at persist.test.js for IE8

### getState method
* Changed to show warning message and not throwing exceptions when storage is polluted and value  comes invalid for persist.
* Use try catch block with session/localStorage when call getState.

### setState method
* If state is null object, make it not call JSON.stringify before store the value.

## Issue
ref #14

## Preferred reviewer
@netil , @kishu

### CODE REVIEW

File: src/plugin/persist.js [hunk 1]
```diff
@@ -71,33 +72,42 @@ eg.module("persist", ["jQuery", eg, window, document], function($, ns, global, d
...
-					if (jQuery.type(state) !== "object" || state instanceof Array) {
-						throw new Error("window.history has no valid format data " +
-							"to be handled in persist.");
-					}
-				} catch (e) {
-					/* jshint ignore:start */
-					console.warn(e.message);
-					/* jshint ignore:end */
-				}
+				isValidStateStr = true;
```
Comment (netil) [inline src/plugin/persist.js]: IMO It doesn't necessarily to do in condition statement.

```js
isValidStateStr = typeof stateStr === "string" && stateStr !== "null";

// line 86
isValidStateStr = stateStr && stateStr.length > 0;
```

Comment (happyhj) [inline src/plugin/persist.js]: I totally agree with you. I will apply it.

**Comment by netil** [comment]
LGTM

**Comment by sculove** [comment]
LGTM

**Pull request closed without merging**